components/cam/src/physics/cam/phys_grid.F90

```diff
@@ -2433,52 +2437,107 @@ subroutine print_cost_p()
+   real(r8) :: avg_ccest              ! average estimated chunk cost
+   real(r8) :: avg_ccost              ! average measured chunk cost
+   real(r8) :: avg_pcest              ! average estimated process cost
+   real(r8) :: avg_pcost              ! average measured process cost
+   real(r8) :: ccost                  ! chunk cost
+   real(r8) :: norm_ccest             ! normalized estimated chunk cost
+   real(r8) :: norm_ccost             ! normalized chunk cost
+   real(r8) :: norm_pcest             ! normalized estimated process cost
+   real(r8) :: norm_pcost             ! normalized process cost
+   real(r8) :: pspeedup               ! local speed-up relative to sum of
+                                      !  local chunk cost, so a measure of
+                                      !  load balance, but not traditional
+                                      !  speed-up (relative to serial cost)
 
    character(len=*), parameter :: fname = 'atm_chunk_costs.txt'
 
 !-----------------------------------------------------------------------
    if (output_chunk_costs) then
       unitn = getunit()
 
-      ! Calculate normalized measured cost
-      cost_gsum = 0.0_r8
-      cost_lsum = 0.0_r8
+      ! Calculate local measured and estimated chunk costs
+      ! and estimated thread and process costs (for bc_physics and
+      ! ac_physics chunk loops). Estimated thread cost is calculated from
+      ! the chunks assigned to each thread, assuming a static round-robin
+      ! distribution. Estimated process cost is then calculated as the
+      ! maximum over the estimated thread costs.
+      ccest_lsum = 0.0_r8
+      ccost_lsum = 0.0_r8
+      tcest(:)   = 0.0_r8
       do lcid = begchunk, endchunk
-         cost_lsum = cost_lsum + lchunks(lcid)%cost
+         ccest_lsum = ccest_lsum + chunks(lchunks(lcid)%cid)%estcost
+         ccost_lsum = ccost_lsum + lchunks(lcid)%cost
+         tid = mod(lcid-begchunk,nlthreads)
+         tcest(tid) = tcest(tid) + chunks(lchunks(lcid)%cid)%estcost
       enddo
+      pcest = maxval(tcest)
+
 #if ( defined SPMD )
-      call MPI_ALLREDUCE(cost_lsum, cost_gsum, 1, MPI_REAL8, MPI_SUM, &
+      gsums(:) = 0.0_r8
+      lsums(1) = ccest_lsum
+      lsums(2) = ccost_lsum
+      lsums(3) = pcest
+      lsums(4) = phys_pcost
+      call MPI_ALLREDUCE(lsums, gsums, 4, MPI_REAL8, MPI_SUM, &
                          mpicom, ierr)
+      ccest_gsum = gsums(1)
+      ccost_gsum = gsums(2)
+      pcest_sum  = gsums(3)
+      pcost_sum  = gsums(4)
 #else
-      cost_gsum = cost_lsum
+      ccost_gsum = ccost_lsum
+      ccest_gsum = ccest_lsum
+      pcost_sum  = phys_pcost
+      pcest_sum  = pcest
 #endif
-      if (abs(cost_gsum) > 0.000001_r8) then
-         avg_cost = cost_gsum/nchunks
-      else
-         avg_cost = 1.0_r8
+
+      ! Calculate average estimated chunk cost
+      avg_ccest = ccest_gsum/real(nchunks,r8)
+      if (avg_ccest < 0.000001_r8) then
+         avg_ccest = -1.0_r8
       endif
```

Comment:
What situation is this checking for? 
All local chunk costs were zero?

Reply:
I hate for code to abort for performance documentation reasons (I prefer erroneous performance data to killing an otherwise working simulation) so put checks like these in to prevent divide-by-zero even if there is not reason for this to ever occur. In this case, perhaps a user specificed that all columns had no cost, in which case the sum was zero? 

The use of the magic number was sloppiness/laziness on my part. There is nothing special about the number, and it should be relative to the numerator. I could just replace it by /= 0.0_r8, to eliminate the one known source of this error, and assume that others simply will never occur? Advice? If we stick to a bound away from a small number, I'll rework this to be more "precise", i.e. < max_ccest*epsilon, or something like this.

Reply:
Reworked  this. Please take a look - google searched for best practice logic to avoid INF and NaN and implemented this - no more magic numbers (and also more robust).

Reply:
Lokks good.